### PR TITLE
feat(ui): enable students to rate tutors

### PR DESCRIPTION
**Pull request by mostafakamar2308**

### CODE REVIEW

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+const RateForm: React.FC<RateFormProps> = ({ tutor, rate }) => {
+  const intl = useFormatMessage();
+
+  const onSuccess = useCallback(() => {
+    toaster.success({ title: "Ratings successful" });
+  }, []);
+  const onError = useCallback((error: Error) => {
+    toaster.error({
+      title: "An error has occured",
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/RateForm.tsx]: Done

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+type RateFormProps = {
+  tutor: number;
+  rate?: IRating.Populated;
+};
+
+const RateForm: React.FC<RateFormProps> = ({ tutor, rate }) => {
+  const intl = useFormatMessage();
+
+  const onSuccess = useCallback(() => {
+    toaster.success({ title: "Ratings successful" });
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/RateForm.tsx]: Done

File: apps/nova/src/components/TutorProfile/EditRating.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Dialog } from "@litespace/luna";
+import React from "react";
+import RateForm from "./RateForm";
+import { IRating } from "@litespace/types";
+
+type EditRatingProps = {
+  title: string;
+  open: boolean;
+  close: () => void;
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/EditRating.tsx:9]:
```suggestion
  close: Void;
```

File: apps/nova/src/components/TutorProfile/EditRating.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { Dialog } from "@litespace/luna";
+import React from "react";
+import RateForm from "./RateForm";
+import { IRating } from "@litespace/types";
+
+type EditRatingProps = {
+  title: string;
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/EditRating.tsx:7]: You don't need to pass the title as a prop. 

```suggestion
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 5]
```diff
@@ -1,39 +1,67 @@
...
+            </Button>
+          </div>
         </div>
         <p>{rating.feedback}</p>
+        <EditRating
+          open={edit.open}
+          close={edit.hide}
+          rate={rating}
+          title={intl("tutor.rate.editmessage")}
+        />
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:64]: Use should only be able to edit the rating if he is the **owner**.

Only show the editing button and the editing dialog incase the current user id is the same as the `rating.raterId` 

You can get the current user from the redux state. 

```
cosnt profile = useAppSelector(profileSelectors.value)

profile.value?.user.id === rating.raterid // <--- user can edit this rating. 
```

File: apps/nova/src/pages/TutorProfile.tsx [hunk 6]
```diff
@@ -86,6 +89,8 @@ const TutorProfile: React.FC = () => {
           <About about={tutor.data.about} />
           <ActivitiesOverview query={activity} />
           <Ratings query={ratings} />
+          {/* New Component to rate tutor */}
```
Comment (neuodev) [inline apps/nova/src/pages/TutorProfile.tsx:92]: Can be deleted. 

```suggestion
```

File: packages/luna/src/locales/ar-eg.json [hunk 7]
```diff
@@ -402,5 +402,8 @@
   "error.schedule.duration": "أقل مده هي 30 دقيقة",
   "error.tutors.list.error": "حدثت مشكلة أثناء تحصيل قائمة المعلمين، الرجاء المحاولة لاحقاً",
   "error.tutors.list.empty": "يبدو أنه لا يوجد معلمين متاحين الان، الرجاء المحاولة لاحقاً",
-  "error.unexpected": "حدث خطاء غير متوقع، الرجاء المحاولة لاحقاً"
+  "error.unexpected": "حدث خطاء غير متوقع، الرجاء المحاولة لاحقاً",
+  "tutor.rate.succes": "تم تقييم المعلم بنجاح",
+  "tutor.rate.error": "حدث خطأ أثناء تقييم المعلم، برجاء المحاولة مرة أخرى",
+  "tutor.rate.editmessage": "تعديل تقييم المعلم"
```
Comment (neuodev) [inline packages/luna/src/locales/ar-eg.json:408]: Please move these message at the top (may be below this [line](https://github.com/litespace-org/litespace/blob/5723ee988380ed9397871cafaeae9bebbc66d0bf/packages/luna/src/locales/ar-eg.json#L34)). The end of the file is reserved for error labels.

**Review by neuodev — CHANGES_REQUESTED** [review]

File: apps/nova/src/components/TutorProfile/EditRating.tsx (hunk 3 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/EditRating.tsx:9]: Alright

File: apps/nova/src/components/TutorProfile/Rating.tsx (hunk 5 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/Rating.tsx:64]: Ok

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      description: error.message,
+    });
+    if (rate && close) {
+      close();
+    }
+  }, []);
+  const rateTutor = useRateTutor({ onSuccess, onError });
+
+  const editRateTutor = useEditRateTutor({
+    id: rate?.id || 0,
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/RateForm.tsx:52]: I am using the id in mutationkey so I need to pass it to the initiation

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 9]
```diff
@@ -1,39 +1,67 @@
...
+import { Edit, Star, Trash2 } from "react-feather";
+import { useRender } from "@/hooks/render";
+import EditRating from "./EditRating";
 
 const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
+  const edit = useRender();
+  const deleteRating = useRender();
+  const intl = useFormatMessage();
+  const showEditDialog = ()=> edit.show();
+  const showDeleteDialog = () => deleteRating.show()
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/Rating.tsx:21]: They are used as event handlers for rating actions

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 10]
```diff
@@ -1,39 +1,67 @@
...
+              <span className="border-2 border-dotted rounded-full border-foreground-light" />
               &nbsp;
               <p>{dayjs(rating.createdAt).fromNow()}</p>
             </p>
           </div>
+          <div className="flex gap-3">
+            <Button onClick={showEditDialog} size={ButtonSize.Tiny} type={ButtonType.Primary}>
+              <Edit className="w-4 text-white" />
+            </Button>
+            <Button onClick={showDeleteDialog} size={ButtonSize.Tiny} type={ButtonType.Primary}>
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/Rating.tsx:53]: Alright, let's go with that.

**Comment by neuodev** [comment]
image

Please squash this commit. And rebase on the latest master and resolve the conflicts.

**Comment by mostafakamar2308** [comment]
@neuodev  
Done

File: apps/nova/src/components/TutorProfile/EditRating.tsx [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import { Dialog, useFormatMessage } from "@litespace/luna";
+import React from "react";
+import RateForm from "./RateForm";
+import { IRating } from "@litespace/types";
+
+type EditRatingProps = {
+  open: boolean;
+  close: () => void;
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/EditRating.tsx:8]: Please use `Void` custom types. 

```suggestion
  close: Void;
```

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    onSuccess,
+    onError,
+  });
+
+  // Create new tutor hook to edit rating
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/RateForm.tsx:55]:
```suggestion
```

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  tutor: number;
+  rate?: IRating.Populated;
+  close?: Void;
+};
+
+const RateForm: React.FC<RateFormProps> = ({ tutor, rate, close }) => {
+  const intl = useFormatMessage();
+  const queryClient = useQueryClient();
+  const onSuccess = useCallback(() => {
+    queryClient.invalidateQueries({ queryKey: ["tutor-rating", tutor] });
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/RateForm.tsx:33]: Please use the new hook `useInvalidate`

File: packages/headless/src/rating.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  });
+}
+
+export function useEditRateTutor({
+  id,
+  onSuccess,
+  onError,
+}: {
+  // id is rating id
+  id: number;
```
Comment (neuodev) [inline packages/headless/src/rating.ts:37]:
```suggestion
```

**Review by neuodev — CHANGES_REQUESTED** [review]

**Comment by neuodev** [comment]
@mostafakamar2308 please don't click "Resolve conversation". Just react to it with 👍. The owner of the comment should resolve it himself.

File: apps/nova/src/components/TutorProfile/RateForm.tsx (hunk 13 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/RateForm.tsx:33]: I added that hook in another pr, do i need to still do this change?

Comment (neuodev) [reply apps/nova/src/components/TutorProfile/RateForm.tsx:33]: no, we can do it later.

**Review by neuodev — APPROVED** [review]
Great work 🚀 @mostafakamar2308! Let's get you first PR merged 🎉

**Pull request merged**